cmd/kube-proxy/app/server.go

```diff
@@ -251,7 +255,8 @@ func (o *Options) initWatcher() error {
 	if err != nil {
 		return err
 	}
-	err = fswatcher.AddWatch(o.ConfigFile)
```

Comment:
do we need to watch both the parent dir and the config file to get all the events we care about? If you run kube-proxy outside a pod and point --config directly at a real file, do we catch all of the following events:
* recreate of the file
* replace of the file
* write to the file

Reply:
Thanks @liggitt 
I don't think this has anything to do with whether or not `kube-proxy` is running in Pod.
We focus on the events of the parent folder, not the event source.
There are only [five fs events](https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/fsnotify/fsnotify/fsnotify.go#L27-L31): `CHMOD`, `REMOVE`, `CREATE`, `RENAME` and `WRITE`, we've dealt with three of them (`CREATE`, `RENAME` and `WRITE`), I think it covers all the possible scenarios. We will decide whether to restart `kube-proxy` by verifying whether the config file has changed and whether the config file contents are valid when the event occurs.

Reply:
does writing/replacing/recreating the config file directly trigger events for a watcher that only registered on the parent dir?

Reply:
> I don't think this has anything to do with whether or not kube-proxy is running in Pod.

When running as a pod, you are exercising the symlink scenario, where an update involves replacing the symlink, which I would expect to trigger events on the parent dir watcher. I want to make sure we also still observe events when the config file itself is directly modified.

Reply:
I've added more test cases, and we should be able to watch all the events we care about.